contrib/drivers/filesharedrivers/chubaofs/chubaofs.go

```diff
@@ -0,0 +1,249 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chubaofs
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+
+	log "github.com/golang/glog"
+	. "github.com/opensds/opensds/contrib/drivers/utils/config"
+	. "github.com/opensds/opensds/pkg/model"
+	pb "github.com/opensds/opensds/pkg/model/proto"
+	"github.com/opensds/opensds/pkg/utils/config"
+	uuid "github.com/satori/go.uuid"
+)
+
+const (
+	DefaultConfPath = "/etc/opensds/driver/chubaofs.yaml"
+	NamePrefix      = "chubaofs"
+)
+
+const (
+	KMountPoint = "mountPoint"
+	KVolumeName = "volName"
+	KMasterAddr = "masterAddr"
+	KLogDir     = "logDir"
+	KWarnLogDir = "warnLogDir"
+	KLogLevel   = "logLevel"
+	KOwner      = "owner"
+	KProfPort   = "profPort"
+)
+
+const (
+	KClientPath = "clientPath"
+)
+
+const (
+	defaultLogLevel = "error"
+	defaultOwner    = "chubaofs"
+	defaultProfPort = "10094"
+
+	defaultVolumeCapLimit int64 = 1000000
+)
+
+const (
+	clientConfigFileName = "client.json"
+	clientCmdName        = "cfs-client"
+)
+
+type ClusterInfo struct {
+	Name           string   `yaml:"name"`
+	MasterAddr     []string `yaml:"masterAddr"`
+	VolumeCapLimit int64    `yaml:"volumeCapLimit"`
+}
+
+type RuntimeEnv struct {
+	MntPoint   string `yaml:"mntPoint"`
+	ClientPath string `yaml:"clientPath"`
+	LogLevel   string `yaml:"logLevel"`
+	Owner      string `yaml:"owner"`
+	ProfPort   string `yaml:"profPort"`
+}
+
+type Config struct {
+	ClusterInfo `yaml:"clusterInfo"`
+	RuntimeEnv  `yaml:"runtimeEnv"`
+
+	Pool map[string]PoolProperties `yaml:"pool,flow"`
+}
+
+type Driver struct {
+	conf *Config
+}
+
+func (d *Driver) Setup() error {
+	conf := &Config{}
+	path := config.CONF.OsdsDock.Backends.Chubaofs.ConfigPath
+	if "" == path {
+		path = DefaultConfPath
+	}
+
+	if _, err := Parse(conf, path); err != nil {
+		return err
+	}
+
+	if conf.MntPoint == "" || conf.ClientPath == "" {
+		return errors.New(fmt.Sprintf("chubaofs: lack of necessary config, mntPoint(%v) clientPath(%v)", conf.MntPoint, conf.ClientPath))
+	}
+
+	if conf.VolumeCapLimit <= 0 {
+		conf.VolumeCapLimit = defaultVolumeCapLimit
+	}
+
+	d.conf = conf
+	return nil
+}
+
+func (d *Driver) Unset() error {
+	return nil
+}
+
+func (d *Driver) CreateFileShare(opt *pb.CreateFileShareOpts) (fshare *FileShareSpec, err error) {
+	log.Info("CreateFileShare ...")
+
+	volName := opt.GetId()
+	volSize := opt.GetSize()
+
+	configFiles, fsMntPoints, owner, err := prepareConfigFiles(d, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	/*
+	 * Only the master raft leader can repsonse to create volume requests.
+	 */
+	leader, err := getClusterInfo(d.conf.MasterAddr[0])
```

Comment:
Is the master address updated by some scheduler/daemon? In case there is new master/leader elected, will it get the updated info?

Reply:
The master address is specified in the driver config file, and will not be changed. So it is either several static ip addresses, or a domain name without ip address predefined. It is according to the on-premise cluster's setup.

Reply:
Ok. Then it's fine. 
I was under the impression that if a master goes down, new leader/master will be elected as per RAFT